From 9eb4967c9328d7700ec0bde790d1147baed39ae6 Mon Sep 17 00:00:00 2001
From: Stephane Viau <sviau@codeaurora.org>
Date: Mon, 13 Jul 2015 15:48:00 -0400
Subject: [PATCH 20/44] drm/msm/mdp5: Disable msm8996 hardware translation
 table walks

On certain targets (eg, MSM8996) we need to set the following
domain attribute for correct operation of the SMMU:
DOMAIN_ATTR_COHERENT_HTW_DISABLE.
Setting that attribute is very important on 8996. Without it, we would
see crazy translation faults.

Change-Id: I1b94d8ee2e92aafbda67628c77adf329b9d1a90c
Signed-off-by: Stephane Viau <sviau@codeaurora.org>
---
 drivers/gpu/drm/msm/mdp/mdp5/mdp5_kms.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/gpu/drm/msm/mdp/mdp5/mdp5_kms.c b/drivers/gpu/drm/msm/mdp/mdp5/mdp5_kms.c
index bbab6e6..14ac20a 100644
--- a/drivers/gpu/drm/msm/mdp/mdp5/mdp5_kms.c
+++ b/drivers/gpu/drm/msm/mdp/mdp5/mdp5_kms.c
@@ -584,6 +584,8 @@ struct msm_kms *mdp5_kms_init(struct drm_device *dev)
 	mdelay(16);
 
 	if (config->platform.iommu) {
+		int rev = mdp5_cfg_get_hw_rev(mdp5_kms->cfg);
+
 		mmu = msm_iommu_new(&pdev->dev, config->platform.iommu);
 		if (IS_ERR(mmu)) {
 			ret = PTR_ERR(mmu);
@@ -591,6 +593,19 @@ struct msm_kms *mdp5_kms_init(struct drm_device *dev)
 			goto fail;
 		}
 
+		/*
+		 * On certain targets (8996, for example) we need to set the
+		 * following domain attribute for correct operation of the SMMU;
+		 * Without it, we would see crazy translation faults..
+		 */
+		if (rev == 7) {
+			int disable_htw = 1;
+
+			iommu_domain_set_attr(config->platform.iommu,
+				DOMAIN_ATTR_COHERENT_HTW_DISABLE, &disable_htw);
+			DBG("coherent hardware translation table walks is off");
+		}
+
 		ret = mmu->funcs->attach(mmu, iommu_ports,
 				ARRAY_SIZE(iommu_ports));
 		if (ret) {
-- 
2.8.2

